File: bindings/dart/lib/src/safe_phone.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe phone number validation via libproven FFI.
///
/// All parsing is performed in the formally verified Idris 2 core.
/// Follows E.164 standard.
library;

import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'ffi.dart';

/// Parsed phone number.
class PhoneNumber {
  /// Country calling code (e.g., 1 for US/CA, 44 for UK).
  final int countryCode;

  /// National number as integer.
  final int nationalNumber;

  const PhoneNumber({
    required this.countryCode,
    required this.nationalNumber,
  });

  /// Format in E.164 format (e.g., +15551234567).
  String? toE164() {
    final result = provenPhoneFormatE164(countryCode, nationalNumber);
    return callStringFfi(result);
  }

  @override
  String toString() => '+$countryCode$nationalNumber';
}

/// Safe phone number operations.
///
/// Delegates to libproven for phone number parsing and formatting.
class SafePhone {
  /// Parse a phone number string.
  /// Returns null on invalid input.
  static PhoneNumber? parse(String input) {
    final (ptr, len) = toNativeUtf8Bytes(input);
    try {
      final result = provenPhoneParse(ptr, len);
      if (result.status != ProvenStatus.ok) return null;
      return PhoneNumber(
        countryCode: result.countryCode,
        nationalNumber: result.nationalNumber,
      );
    } finally {
      calloc.free(ptr);
    }
  }

  /// Check if a phone number string is valid.
  static bool isValid(String input) {
    return parse(input) != null;
  }

  /// Format a phone number in E.164 format.
  /// Returns null on invalid input.
  static String? formatE164(String input) {
    final phone = parse(input);
    return phone?.toE164();
  }
}

File: bindings/rust/src/safe_phone.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe phone number operations via libproven FFI.
//!
//! Parses and formats E.164 phone numbers.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Parsed phone number.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PhoneNumber {
    /// Country calling code (e.g., 1 for US, 44 for UK).
    pub country_code: u16,
    /// National number (without country code).
    pub national_number: u64,
    /// Whether the phone number passed validation.
    pub is_valid: bool,
}

/// Safe phone number operations.
pub struct SafePhone;

impl SafePhone {
    /// Parse a phone number string.
    ///
    /// Accepts various formats including E.164 (`+1234567890`).
    pub fn parse(input: &str) -> Result<PhoneNumber> {
        let bytes = input.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        // The FFI function only reads from the pointer and returns a
        // value-type struct.
        let result = unsafe {
            ffi::proven_phone_parse(bytes.as_ptr(), bytes.len())
        };
        core::status_to_result(result.status)?;
        Ok(PhoneNumber {
            country_code: result.country_code,
            national_number: result.national_number,
            is_valid: result.is_valid,
        })
    }

    /// Format a phone number as E.164 (e.g., "+12025551234").
    pub fn format_e164(country_code: u16, national_number: u64) -> Result<String> {
        // SAFETY: proven_phone_format_e164 takes value-type arguments;
        // always safe to call.
        let result = unsafe {
            ffi::proven_phone_format_e164(country_code, national_number)
        };
        core::string_result_to_result(result)
    }
}
